controllers/dataprotection/backup_controller.go

```diff
@@ -870,26 +908,19 @@ func getClusterObjectString(cluster *appsv1alpha1.Cluster) (*string, error) {
 			constant.ExtraEnvAnnotationKey: v,
 		}
 	}
-	clusterBytes, err := json.Marshal(newCluster)
-	if err != nil {
-		return nil, err
-	}
-	clusterString := string(clusterBytes)
-	return &clusterString, nil
+	clusterString, err := getObjectString(newCluster)
+	return clusterString, err
 }
 
-// setClusterSnapshotAnnotation sets the snapshot of cluster to the backup's annotations.
-func setClusterSnapshotAnnotation(backup *dpv1alpha1.Backup, cluster *appsv1alpha1.Cluster) error {
+// setSnapshotAnnotation sets the snapshot of cluster to the backup's annotations.
+func setSnapshotAnnotation(request *dpbackup.Request, cluster *appsv1alpha1.Cluster) error {
+	if request.Backup.Annotations == nil {
+		request.Backup.Annotations = map[string]string{}
+	}
 	clusterString, err := getClusterObjectString(cluster)
-	if err != nil {
+	if err != nil || clusterString == nil {
```

Comment:
It's too implicit that the function returns nil when `clusterString == nil`. I think the original code is easier to understand.